bsp: Add display and touch support to gerenic BSP.

espzav opened the pull request:
# ESP-BSP Pull Request checklist

- [x] Version of modified component bumped
- [x] CI passing

# Change description
- Updated `led_indicator` and support classic RGB LED (e.g. ESP32C2)
- Added LCD and touch support to generig BSP. I selected only some drivers
- LCD is supported only via SPI
- Touch is supported only via I2C

# Issues
- When some part not used (e.g. buttons, leds, LCD, touch), the component is still downloaded
- Components for all touches and LCDs are downloaded

tore-espressif commented on bsp/esp_bsp_generic/src/esp_bsp_generic.c:
```diff
@@ -374,6 +416,288 @@ esp_err_t bsp_sdcard_unmount(void)
+
+    ESP_LOGD(TAG, "Install LCD driver");
+    const esp_lcd_panel_dev_config_t panel_config = {
+        .reset_gpio_num = BSP_LCD_RST, // Shared with Touch reset
```
Nitpick:
```suggestion
        .reset_gpio_num = BSP_LCD_RST,
```

tore-espressif commented on bsp/esp_bsp_generic/Kconfig:
```diff
@@ -334,11 +334,14 @@ menu "Board Support Package"
             
         choice  
             prompt "LED type"
+            depends on BSP_BUTTONS_NUM > 0
```
Why does 'LED type' depend on `BSP_BUTTONS_NUM > 0` ?

espzav replied:
Copy paste mistake. Thanks.

tore-espressif reviewed (commented):
* There is a missing comma at `esp_bsp_generic.c:77` and similarly below. Button ADC fails to compile
* It is a little misleading that I can still set 'Number of LEDs' to 5, but I can define only 1 'Classic RGB LED'. Can we hide the 'number of LEDs' menu for classic RGB LEDs?

espzav commented on bsp/esp_bsp_generic/include/bsp/display.h:
```diff
@@ -0,0 +1,72 @@
+/* LCD display color bits */
+#define BSP_LCD_BITS_PER_PIXEL      (16)
+/* LCD display color space */
+#define BSP_LCD_COLOR_SPACE         (ESP_LCD_COLOR_SPACE_BGR)
```
`BSP_LCD_COLOR_SPACE` was added to KConfig.

espzav commented:
> * There is a missing comma at `esp_bsp_generic.c:77` and similarly below. Button ADC fails to compile
> * It is a little misleading that I can still set 'Number of LEDs' to 5, but I can define only 1 'Classic RGB LED'. Can we hide the 'number of LEDs' menu for classic RGB LEDs?

@tore-espressif Thank you for review! Fixed all mistakes and changed LEDs in KConfig - the first select type and then number of leds (zero still means no leds).

tore-espressif commented on bsp/esp_bsp_generic/Kconfig:
```diff
@@ -326,38 +326,87 @@ menu "Board Support Package"
+                int
+                default 5
+        endif
+        if BSP_LED_TYPE_RGB_CLASSIC || BSP_LED_TYPE_RGB
```
Should `ENV_MAX_LEDS` be very large for `BSP_LED_TYPE_RGB`? That means a very long led_strip?

With current settings the addressable RGB leds are limited to 1

espzav replied:
Because adressable RGB led is set as one, but you can control it by index in `led_indicator`.

tore-espressif commented on bsp/esp_bsp_generic/Kconfig:
```diff
@@ -479,7 +518,283 @@ menu "Board Support Package"
+                bool "BGR"
+        endchoice
+
+        config BSP_DISPLAY_CMD_BITS
+            depends on BSP_DISPLAY_ENABLED
+            int "LCD command bits"
+            default 8
+            help
+                The command bits of the display.
+
+        config BSP_DISPLAY_PARAM_BITS
+            depends on BSP_DISPLAY_ENABLED
+            int "LCD param bits"
+            default 8
+            help
+                The param bits of the display.
```
This should be constant for given LCD controller. Eg. for ILI9341 this is always 8, right?

tore-espressif replied:
Sorry, I did not write it clearly. Could we hide this config option?

espzav replied:
Sure. Done.

tore-espressif commented on bsp/esp_bsp_generic/Kconfig:
```diff
@@ -479,7 +518,283 @@ menu "Board Support Package"
+            config BSP_DISPLAY_BACKLIGHT_GPIO
+                int 
+                prompt "Display backlight GPIO"
+                default 0
```
default -1?

tore-espressif commented on bsp/esp_bsp_generic/Kconfig:
```diff
@@ -479,7 +518,283 @@ menu "Board Support Package"
+                
+        config BSP_DISPLAY_BRIGHTNESS_LEDC_CH
+            depends on BSP_DISPLAY_ENABLED
+            int "LEDC channel index"
```
Should be visible only if `BSP_DISPLAY_BACKLIGHT_GPIO` != -1 ?
```suggestion
            int "LEDC channel index for brightness control"
```

tore-espressif replied:
I have a board that have inverted logic of the brightness control. Can we add this option too?

espzav replied:
Good point! Done.

tore-espressif reviewed (commented):
@espzav Thanks for the updates, I left few comments worth considering.

I will test with a real custom board and LCD tomorrow

espzav commented:
@tore-espressif Thank you for the review. Fixed and one comment left.

tore-espressif commented on bsp/esp_bsp_generic/Kconfig:
```diff
@@ -479,7 +518,285 @@ menu "Board Support Package"
+            default 100
+            range 10 240
+            help
+                Framebuf is used for lvgl rendering output.
```
Could you please add information about units of this value? eg. row, bytes...

espzav replied:
added

tore-espressif commented on bsp/esp_bsp_generic/src/esp_bsp_generic.c:
```diff
@@ -374,6 +416,289 @@ esp_err_t bsp_sdcard_unmount(void)
+#endif
+    esp_lcd_panel_reset(*ret_panel);
+    esp_lcd_panel_init(*ret_panel);
+    esp_lcd_panel_mirror(*ret_panel, true, true);
```
Why do we mirror the LCD regardless of users configuraion?

espzav replied:
Sorry, I missed it. Reworked. Thank you.

tore-espressif commented on bsp/esp_bsp_generic/src/esp_bsp_generic.c:
```diff
@@ -374,6 +416,289 @@ esp_err_t bsp_sdcard_unmount(void)
+#endif
+    esp_lcd_panel_reset(*ret_panel);
+    esp_lcd_panel_init(*ret_panel);
+    esp_lcd_panel_mirror(*ret_panel, true, true);
```
Some displays also need color inversion here

espzav replied:
Added.

tore-espressif reviewed (commented):
@espzav Thanks for the updates. I tested with ESP32-S3-EYE and found few things that did not work out of the box

espzav commented:
@tore-espressif Thank you for the review and HW test. Everything fixed/added.
